replace dash with underscore in os name

scheibelp opened the pull request:
Potential fix for https://github.com/spack/spack/issues/7356

It appears that `OperatingSystem` should be careful to format the provided name, since it looks like that name may contain a `-` (which is used by `ArchSpec` to delineate between platform, os, and target), for example if the `distname` reported by `external.distro.linux_distribution` (as used in `LinuxDistro`) contains a `-` (I should say I'm not 100% sure if `external.distro.linux_distribution` allows `-` but a cursory glance suggests it may be possible).

healther commented:
This doesn't seem to have any effect (at all), which I find very strange.

```
<checkout this pr>
$ rm -r ~/.spack/
$ spack install zlib
==> Error: linux-debianstable-updates-x86_64 is an invalid arch spec
$ git log --oneline
4427234dc (HEAD -> PR/archnamefix) replace dash with underscore in os name
189d9ec13 Rsync package: Fix Src Link (#7367)
6213463ae new package: hiop (#7039)
86afd0e85 new package: perl-parallel-forkmanager (#7049)
1ac0bce3e help gdb pick up python3 (#7104)
b4b01b82a mpich@develop: needs git submodules (#7357)
da2d6cccd new package: py-goatools (#7328)
d3ed46e95 [spack/spec.py] raise a query failure error if a property query returns None (#7277)
98949bd79 change recurse into recursive in a few places (#7342)
3089207ce minighost: fix build on Ubuntu (#7354)
6c2fe27f4 examinimd: fix a typo (#7352)
f302d6423 (origin/develop, origin/HEAD, develop) add package verilator (#7332)
```

scheibelp commented:
@healther I think I was off and the `-` was actually in the version, so the latest commit converts that as well. I read `LinuxDistro` more carefully and it actually allows `-` in the version name:

```
>>> z = 'hi-there fellow'
>>> re.split(r'[^\w-]', z)
['hi-there', 'fellow']
>>> re.split(r'\W', z)
['hi', 'there', 'fellow']
```

@tgamblin is there a reason you prefer to allow the version components to contain a `-`? If so then I think it's reasonable to do the conversion in `OperatingSystem.__init__` as in https://github.com/spack/spack/pull/7381/commits/39491786573ea96909822eb041c60134b49dccaf

healther commented:
This one works and results in the following in compilers.yaml:
```
operating_system: debianstable_updates
```

healther commented:
can we merge this and close #7356?
